components/brave_adaptive_captcha/brave_adaptive_captcha_service.cc

```diff
@@ -0,0 +1,189 @@
+  const std::string captcha_id = prefs_->GetString(kScheduledCaptchaId);
+  if (payment_id.empty() || captcha_id.empty()) {
+    return false;
+  }
+
+  const int failed_attempts =
+      prefs_->GetInteger(kScheduledCaptchaFailedAttempts);
+
+  *url = brave_adaptive_captcha::BraveAdaptiveCaptchaService::
+      GetScheduledCaptchaUrl(payment_id, captcha_id);
+  *max_attempts_exceeded =
+      failed_attempts >= kScheduledCaptchaMaxFailedAttempts;
+
+  return true;
+}
+
+void BraveAdaptiveCaptchaService::UpdateScheduledCaptchaResult(bool result) {
+  if (!result) {
+    const int failed_attempts =
+        prefs_->GetInteger(kScheduledCaptchaFailedAttempts) + 1;
+    prefs_->SetInteger(
+        kScheduledCaptchaFailedAttempts,
+        std::min(failed_attempts, kScheduledCaptchaMaxFailedAttempts));
+    if (failed_attempts >= kScheduledCaptchaMaxFailedAttempts) {
+      prefs_->SetBoolean(kScheduledCaptchaPaused, true);
+    }
+    return;
+  }
+
+  ClearScheduledCaptcha();
+}
+
+void BraveAdaptiveCaptchaService::ShowScheduledCaptcha(
+    const std::string& payment_id,
+    const std::string& captcha_id) {
+  if (prefs_->GetBoolean(kScheduledCaptchaPaused)) {
+    return;
+  }
+
+  prefs_->SetString(kScheduledCaptchaPaymentId, payment_id);
+  prefs_->SetString(kScheduledCaptchaId, captcha_id);
+
+  if (delegate_) {
+    delegate_->ShowScheduledCaptcha(payment_id, captcha_id);
+    return;
+  }
+}
+
+void BraveAdaptiveCaptchaService::SnoozeScheduledCaptcha() {
+  const int snooze_count = prefs_->GetInteger(kScheduledCaptchaSnoozeCount);
+  if (snooze_count >= 1) {
+    return;
+  }
+
+  prefs_->SetString(kScheduledCaptchaPaymentId, "");
+  prefs_->SetString(kScheduledCaptchaId, "");
+  prefs_->SetInteger(kScheduledCaptchaSnoozeCount, snooze_count + 1);
+}
+
+void BraveAdaptiveCaptchaService::ClearScheduledCaptcha() {
+  prefs_->SetInteger(kScheduledCaptchaFailedAttempts, 0);
+  prefs_->SetInteger(kScheduledCaptchaSnoozeCount, 0);
+  prefs_->SetString(kScheduledCaptchaPaymentId, "");
+  prefs_->SetString(kScheduledCaptchaId, "");
+  prefs_->SetBoolean(kScheduledCaptchaPaused, false);
+}
+
+void BraveAdaptiveCaptchaService::OnRecoverWallet(
+    brave_rewards::RewardsService* rewards_service,
+    const ledger::type::Result result) {
+  ClearScheduledCaptcha();
+}
+
+void BraveAdaptiveCaptchaService::OnCompleteReset(const bool success) {
+  ClearScheduledCaptcha();
+}
+
+// static
+void BraveAdaptiveCaptchaService::RegisterProfilePrefs(
+    PrefRegistrySimple* registry) {
+  registry->RegisterStringPref(kScheduledCaptchaId, "");
+  registry->RegisterStringPref(kScheduledCaptchaPaymentId, "");
+  registry->RegisterIntegerPref(kScheduledCaptchaSnoozeCount, 0);
+  registry->RegisterIntegerPref(kScheduledCaptchaFailedAttempts, 0);
+  registry->RegisterBooleanPref(kScheduledCaptchaPaused, false);
+}
+
+// static
+std::string BraveAdaptiveCaptchaService::GetScheduledCaptchaUrl(
```

Comment:
why isn't this in an anonymous namespace? I can't see any reason to make it a public static method

Reply:
Yep, that fell out of refactoring. Fixed.